### ISSUE TITLE
NuGet/Dotnet restore leaves files behind in /tmp/NuGetScratch

### ISSUE THREAD

**Issue by lbussell** [NuGet/Home#12420]
### NuGet Product Used

dotnet.exe

### Product Version

Latest version of .NET 6.0, .NET 7.0, and pre-release versions of .NET 8.0

### Worked before?

_No response_

### Impact

It's more difficult to complete my work

### Repro Steps & Context

1. Build the following Dockerfile (ex. `docker build -t nuget-test -f .\Dockerfile .`)

```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /source
RUN dotnet new console --no-restore
RUN dotnet restore --use-current-runtime
```

2. Run the created image: `docker run -it --rm nuget-test /bin/sh`
3. Inspect the `/tmp` directory to see the files left behind: `ls -lah /tmp/NuGetScratch/lock/`

Is this the expected result? These files left behind by NuGet are causing us permission issues when switching between root and non-root users in .NET docker containers. When a NuGet/dotnet restore is run as root, future NuGet commands ran by other users will fail. Is there a build server of sort that we accidentally terminate early in the container build?

Thanks.

### Verbose Logs
The following are the files left behind by NuGet in an example image that I built using the instructions above (some of the files are not due to NuGet restore like `.dotnet/shm`, while others like MSBuildTemproot are left behind by powershell installation):

```
drwxrwxrwx         0:0        0 B  ├── tmp                                                                                                                                                                                                                                    drwxrwxrwx         0:0        0 B  │   ├── .dotnet                                                                                     drwxrwxrwx         0:0        0 B  │   │   └── shm                                                                                         drwx------         0:0        0 B  │   ├── 82d9253f-fac5-4dd9-863e-e95ad886c6f0                                                            drwx------         0:0        0 B  │   ├── MSBuildTemproot                                                                                 drwxrwxrwx         0:0        0 B  │   ├── NuGetScratch                                                                                    drwxrwxrwx         0:0        0 B  │   │   └── lock                                                                                        -rw-r--r--         0:0        0 B  │   │       ├── 02c1a6b1e6a189adc09794fab6791bf24a1841dd                                                -rw-r--r--         0:0        0 B  │   │       ├── 3b3fba4ea8f8b14f53e82a125bf28c4a69896605                                                -rw-r--r--         0:0        0 B  │   │       ├── 6f70e9594edbb2d12174c395fa5596b59259ef3e                                                -rw-r--r--         0:0        0 B  │   │       ├── 910ee62b266dc259a5a66981bebe7d3cd7cc2a82                                                -rw-r--r--         0:0        0 B  │   │       ├── ad41768f5ec8770e070473ab153efe7a828bba73                                                -rw-r--r--         0:0        0 B  │   │       ├── b5470a0e371577b0df3c0b80c7c2bc6849cd2ca6                                                -rw-r--r--         0:0        0 B  │   │       ├── b5f9502d257d6ab45cab231e13e8c7dd39c7d3a4                                                -rw-r--r--         0:0        0 B  │   │       ├── ed9ecc1f06bc6d3cb9e44cb780575b171eaf893e                                                -rw-r--r--         0:0        0 B  │   │       ├── effbb02e5017975996b2720d0dde89125b10edcc                                                -rw-r--r--         0:0        0 B  │   │       ├── f1cb54aa7ea80f31ab84d0bf6120755929641ad2                                                -rw-r--r--         0:0        0 B  │   │       └── ff442a5005e856d0388cb8daf4c3340cc673f37a                                                drwx------         0:0        0 B  │   ├── c8c62065-522d-429f-b05b-2e52c3084165                                                            drwx------         0:0        0 B  │   └── de16094a-128e-48a0-8ae3-033310547545
```

(The logs don't seem to show up correctly - maybe inspect the raw post to see them)

**Comment by lbussell** [comment]
cc @mthalman

**Comment by heng-liu** [comment]
Hi @lbussell , thanks for filing this issue!
I could repro this issue. This issue is under consideration.
For now, could you try the following workarounds and see if any one works for you? Thanks!

1.Run `dotnet nuget locals temp -c` to clear the /tmp/NuGetScratch   (If there is a sticky bit in /tmp permissions then it won't work)

2.Set environment variable `NUGET_ConcurrencyUtils_DeleteOnClose` to 1 before running restore, so the lock files will be cleared after restore (This [change](https://github.com/NuGet/NuGet.Client/commit/b56eacca7b4d73463631ec82c4b52867c13482b2) is only available in NuGet 6.2 and above, probably .NET 7 preview6 and above)

3.Set environment variable NUGET_SCRATCH to a path. This variable will override the default temp folder. But it's only applied to NuGet version 6.2 and later, probably .NET 7 preview6 and above).

**Comment by Nefcanto** [comment]
This has paralyzed us for almost two days. We use **Dev Containers** extension to attach to a container that has all of the dependencies cached. And we constantly destroy and create containers. Each time it takes us a couple of minutes to restore files and many times we see `lock` exception.

Please fix this sooner. It's very frustrating. Thank you.

**Comment by Nefcanto** [comment]
This has paralyzed us for almost two days. We use **Dev Containers** extension to attach to a container that has all of the dependencies cached. And we constantly destroy and create containers. Each time it takes us a couple of minutes to restore files and many times we see `lock` exception.

Please fix this sooner. It's very frustrating. Thank you.

**Comment by TheDevelolper** [comment]
I am having the same issue pretty much a year on! I'm using Dev Containers and it creates and leaves all these folders in place?!

**Comment by StachuDotNet** [comment]
> Hi @lbussell , thanks for filing this issue! I could repro this issue. This issue is under consideration. For now, could you try the following workarounds and see if any one works for you? Thanks!
> 
> 1.Run `dotnet nuget locals temp -c` to clear the /tmp/NuGetScratch (If there is a sticky bit in /tmp permissions then it won't work)
> 
> 2.Set environment variable `NUGET_ConcurrencyUtils_DeleteOnClose` to 1 before running restore, so the lock files will be cleared after restore (This [change](https://github.com/NuGet/NuGet.Client/commit/b56eacca7b4d73463631ec82c4b52867c13482b2) is only available in NuGet 6.2 and above, probably .NET 7 preview6 and above)
> 
> 3.Set environment variable NUGET_SCRATCH to a path. This variable will override the default temp folder. But it's only applied to NuGet version 6.2 and later, probably .NET 7 preview6 and above).

I just faced this today, on linux (ubuntu 22.04), on a fresh install of Ubuntu/Docker/VSCode/etc, in a devcontainer.

For some reason, in my case, NuGet tried to use an incorrect directory - `/tmp/NuGetScratch[myProjectName]`, rather than just `/tmp/NuGetScratch`. Setting the env var `NUGET_SCRATCH` to `/tmp/NuGetScratch` resolved the issue. (so, I guess it tried to store it under one dir, and fetch it via another? not sure.) Doesn't really answer why this problem is popping up where it didn't before, but at least it unblocks me.

**Comment by joecarl** [comment]
Im facing the same issue. Sometimes on restoring I receive the following error: `Error setting permissions to directory /tmp/NuGetScratchvscode`.
I will try setting the env var `NUGET_SCRATCH` to `/tmp/NuGetScratch` as @StachuDotNet says... but not sure this will solve it forever.
